test/integration_test.go

```diff
@@ -0,0 +1,260 @@
+// panic.
+func getWoodpeckerMetrics(host string) string {
+	url := fmt.Sprintf("%s/metrics", host)
+
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		panic(fmt.Sprintf("Unable to make GET request: %s",
+			err.Error()))
+	}
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		panic(fmt.Sprintf("Unable to get ct-woodpecker metrics with GET to %q: %s",
+			url, err.Error()))
+	}
+	if resp.StatusCode != http.StatusOK {
+		panic(fmt.Sprintf("Unable to get ct-woodpecker metrics with GET to %q: response status %d",
+			url, resp.StatusCode))
+	}
+
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		panic(fmt.Sprintf("Unable to read GET response body: %s", err.Error()))
+	}
+
+	return string(body)
+}
+
+// safeBuffer is a wrapper around a bytes.Buffer that is made safe for
+// concurrent access. This is required for making a logger backed by a bytes
+// buffer that can be used by a woodpecker instance across multiple goroutines
+// without a data race.
+type safeBuffer struct {
+	b bytes.Buffer
+	m sync.RWMutex
+}
+
+func (b *safeBuffer) Read(p []byte) (n int, err error) {
+	b.m.RLock()
+	defer b.m.RUnlock()
+	return b.b.Read(p)
+}
+func (b *safeBuffer) Write(p []byte) (n int, err error) {
+	b.m.Lock()
+	defer b.m.Unlock()
+	return b.b.Write(p)
+}
+
+func (b *safeBuffer) String() string {
+	b.m.RLock()
+	defer b.m.RUnlock()
+	return b.b.String()
+}
+
+// wodpeckerRun starts a ct-woodpecker monitor with the given configuration. It
+// is allowed to run for the given number of fetchIterations. When complete the
+// standard out and the raw metrics data from the woodpecker instance are
+// returned. If there are any unexpected problems an error is returned instead.
+func woodpeckerRun(conf woodpecker.Config, fetchIterations int) (string, string, error) {
+	// Create a logger backed by the safeBuffer. The log.Logger type is only safe
+	// for concurrent use when the backing buffer is. Using a raw bytes.Buffer
+	// with a shared logger will cause data races.
+	var out safeBuffer
+	logger := log.New(&out, "integration test ", log.LstdFlags)
+	clk := clock.Default()
+
+	// Create a Woodpecker with the provided monitoring configuration
+	woodpecker, err := woodpecker.New(conf, logger, clk)
+	if err != nil {
+		return "", "", err
+	}
+
+	// Calculate the fetchInterval duration from the configuration
+	fetchInterval, err := time.ParseDuration(conf.STHFetchInterval)
+	if err != nil {
+		return "", "", err
+	}
+
+	// Start the monitoring process
+	woodpecker.Run()
+
+	// Sleep for the right amount of time based on the fetchInterval and the
+	// requested number of iterations
+	time.Sleep(fetchInterval * time.Duration(fetchIterations))
+
+	// Collect metrics from the woodpecker instance while it is still running
+	metricsData := getWoodpeckerMetrics("http://localhost:1971")
+
+	// Stop the monitoring process
```

Comment:
I think this is an example of "too much comment." The line below `woodpecker.Stop()` is pretty self-explanatory.

Reply:
Guilty as charged. In my defence I think I wrote these comments before I had the implementation and used them as a skeleton. Should have deleted them after :-)